### ISSUE TITLE
PIN policy enhancements

### ISSUE THREAD

**Issue by quynhnng** [privacyidea/privacyidea#579]
# Versions

privacyIDEA: 2.16.1

Installation method:

- from source / github

## more details: 

OS: amazon linux

Webserver:

Tokendatabase: PostgreSQL

=====

Hi Cornelius,

There are several enhancement regarding PIN policies that we think are important:

**1. Check PIN policy when assigning a token:**

Currently, PrivacyIDEA only check PIN policy during enrollment. It is important that we also check during assignment as user can also input the PIN at this point.
For example, we was able to update In privacyidea/api/token.py with the below to enforce for otp PIN check:
```
340 @token_blueprint.route('/assign', methods=['POST'])
341 @prepolicy(check_max_token_realm, request)
342 @prepolicy(check_max_token_user, request)
343 @prepolicy(check_base_action, request, action=ACTION.ASSIGN)
344 @prepolicy(encrypt_pin, request)
+++ @prepolicy(check_otp_pin, request)
345 @prepolicy(check_external, request, action="assign")
346 @event("token_assign", request, g)
347 @log_with(log)
348 def assign_api():
```

**2. Add PIN policy for SPASS token:**

We are using SPASS token for temporary and emergency cases, where other token types are not applicable. However, SPASS token does not currently have a configurable PIN policy, and can be set to a weak PIN or even left empty. This opens security risk to allow this token, and highly effects its usability.

SPASS token should have PIN policy configurable similar to OTP PIN policy: min/max length, and required contents.

**3. PIN policy applicable for admin scope:**

Currently, PIN policy actions are only available to 'user' scope. It would be nice to also give an option to enforce it in admin scope for fool proof.

 On the other notes, when go through the policy, I also notice a tiny typo:
In privacyidea/api/lib/prepolicy.py:
```
247            raise PolicyError("The maximum OTP PIN length is {0!s}".format(
248                              pol_minlen[0]))
```
where `pol_minlen[0]` should be ` pol_maxlen[0]`

Thank you for your help, and I am looking forward to your feed back!

**Comment by cornelinux** [comment]
**1. Check PIN policy when assigning a token:**

OK, we would have to check, if there are any side effects, to other use cases. But this looks straight forward.

Regarding your 2nd and 3rd point I would prefer to first discuss your workflows.
There are other possibilities (lost tokens) for emergency facilitating the "password" token.

**Comment by cornelinux** [comment]
I just checked a user enrolling SPASS token. The pin policy in fact is applied. So No. 2 is no issue. But I think and administrator enrolled a SPASS token. So the No. 2 also boils down to No. 3.

**Comment by axiadidsProdops** [comment]
Thanks for the feedback Cornelius. Regarding 2. you are right the pin policy is applied, however we would need to be able to decouple the Password policy for PIN from the one for SPASS. Since the use case is different. 

PIN are usually simple (numerical only - 4-8 characters)
SPASS used for emergency/temporary access should be more complex and have their own policy.

Thanks

**Comment by cornelinux** [comment]
Decoupling is not possible. You should take a look at the lost token scenario. We really should discuss your workflows.
SPASS is not ment for emergency or temporary access. The "lost token" is.
From a design standpoint there is no sense in defining special SPASS pin policies.

**Comment by quynhnng** [comment]
In our senerio, we are using SPASS token for temporary token for those that are waiting for their hardware token. Our infrastructure only allows hardware token assigned by user user, and SPASS token is enrolled by operators.

**Comment by quynh-axiadids** [comment]
Hi Cornelinux,

As this is of urgent, I have the following modifications to allow a dedicate SPASS policy to be enforced, and would like to ask for your feedback:

[https://github.com/axiadids/privacyidea/commit/8216c45872da1185c160d394cd175c8956951686](url)
[https://github.com/axiadids/privacyidea/commit/479fde827b2984ffc41042deef505c9116bd016d](url)

Thank you very much.

**Comment by quynh-axiadids** [comment]
Regards admin scope for PIN policy, I have also attempted to extend PIN policy to be able to enforce for admin role with this commit:
96de268070fd78bc521e3f95662d4dba46b0ba96

I have tested with my environment, and it looks good. I am looking to your feedback on allow privacyidea this feature.

**Comment by cornelinux** [comment]
Hi Quynh,
I added some comments.
Basically I would not like to add special SPASS PIN policies.
Adding PIN policies for scope=admin looks sensible to me.
Kind regards
Cornelius

**Comment by quynh-axiadids** [comment]
Thank you for your feedbacks!

I have updated and cleaned up the commits as you suggested as whole new commits:
- Admin scope PIN policy: ef76bb790b2ef81ab5ccb75a12cd0dbf3d6c2596
- Spass policy: 6ffb4a9c190f1ee5ff985527fbc7ec28ae804a42

For the SPASS extra decorator, I decided to keep the modification within the same old call due to code repetition and possible PIN policy conflicts/decisions. They are all otp-pin related policies after all in my opinion. But maybe it can be improved for the better.

We are looking forward to discuss with you next week!
